crates/oxc_linter/src/config/flat.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    pub(crate) fn resolve(&self, path: &Path) -> ResolvedLinterState {
+        if self.overrides.is_empty() {
+            return self.base.clone();
+        }
+
+        // SAFETY: number of overrides is checked in constructor, and overrides cannot be added
+        // after ConfigStore is created.
+        unsafe { assert_unchecked!(self.overrides.len() <= MAX_OVERRIDE_COUNT) };
+        // Resolution gets run in a relatively tight loop. This vec allocates on the stack, kind
+        // of like `int buf[MAX_OVERRIDE_COUNT]` in C (but we also keep track of a len). This
+        // prevents lots of malloc/free calls, reducing heap fragmentation and system call
+        // overhead.
+        let mut overrides_to_apply: StackVec<OverrideId, MAX_OVERRIDE_COUNT> = StackVec::new();
+        let mut hasher = FxBuildHasher.build_hasher();
+
+        for (id, override_config) in self.overrides.iter_enumerated() {
+            // SAFETY: we know that overrides_to_apply's length will always be less than or equal
+            // to the maximum override count, which is what the capacity is set to. This assertion
+            // helps rustc optimize away bounds checks in the loop's `.push()` calls. Rustc is
+            // notoriously bad at optimizing away loop bounds checks, so we do it for it.
+            unsafe { assert_unchecked!(overrides_to_apply.len() < overrides_to_apply.capacity()) };
+            if override_config.files.is_match(path) {
+                overrides_to_apply.push(id).unwrap();
+                id.hash(&mut hasher);
+            }
+        }
+
+        if overrides_to_apply.is_empty() {
+            return self.base.clone();
+        }
+
+        let key = hasher.finish();
+        self.cache
+            .entry(key)
+            .or_insert_with(|| self.apply_overrides(&overrides_to_apply))
+            .value()
+            .clone()
+    }
+
+    /// NOTE: this function must not borrow any entries from `self.cache` or DashMap will deadlock.
+    fn apply_overrides(&self, override_ids: &[OverrideId]) -> ResolvedLinterState {
+        let plugins = self
+            .overrides
+            .iter()
+            .rev()
+            .find_map(|cfg| cfg.plugins)
+            .unwrap_or(self.base.config.plugins);
+
+        let all_rules = RULES
+            .iter()
+            .filter(|rule| plugins.contains(LintPlugins::from(rule.plugin_name())))
+            .cloned()
+            .collect::<Vec<_>>();
+        let mut rules = self
+            .base
+            .rules
+            .iter()
+            .filter(|rule| plugins.contains(LintPlugins::from(rule.plugin_name())))
+            .cloned()
+            .collect::<FxHashSet<_>>();
+
+        let overrides = override_ids.iter().map(|id| &self.overrides[*id]);
+        for override_config in overrides {
+            if override_config.rules.is_empty() {
+                continue;
+            }
+            override_config.rules.override_rules(&mut rules, &all_rules);
```

Comment:
We could pass `RULES` directly here, avoiding the need to allocate another Vec. It is going to be larger though, so the lookup is slower. I profiled locally and there was no noticeable perf difference though:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/50a761a4-23d9-41f1-a402-c2937be02a89">

In the future, we might want to consider making this a hash set to make the lookup faster?

```suggestion
            override_config.rules.override_rules(&mut rules, &RULES);
```